File: src/javascript/algorithms/linkedlist/6.reorderList/reorderList.js
 // TC: O(n) SC: O(1)

function reorderList(head) {
    if(head === null) return;
    
    //Find middle
    let slow = head, fast = head;
    while(fast.next != null && fast.next.next != null) {
        slow = slow.next;
        fast = fast.next.next;
    }

    //Reverse the second half of linkedlist
    let secondHalf = slow.next;
    let prev = slow.next = null;
    while(secondHalf != null) {
        let temp = secondHalf.next;
        secondHalf.next = prev;
        prev = secondHalf;
        secondHalf = temp;
    }

    //Re-order the full list
    let firstHalf = head;
    secondHalf = prev;
    while(secondHalf != null) {
        let temp1 = firstHalf.next, temp2 = secondHalf.next;
        firstHalf.next = secondHalf;
        secondHalf.next = temp1;
        firstHalf = temp1;
        secondHalf = temp2;
    }
}

function printLinkedList(head) {
    let temp = head;
    while(temp != null) {
        console.log(temp.value + " ");
        temp = temp.next;
    }
}

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }

    setNext(next) {
        this.next = next;
    }
}

let headNode1 = new Node(1);
let node2 = new Node(2);
let node3 = new Node(3);
let node4 = new Node(4);
let node5 = new Node(5);
let node6 = new Node(6);

headNode1.setNext(node2);
node2.setNext(node3);
node3.setNext(node4);
node4.setNext(node5);
node5.setNext(node6);

let headNode2 = null;

reorderList(headNode1)
printLinkedList(headNode1);

reorderList(headNode2)
printLinkedList(headNode2);

File: src/python/algorithms/linkedlist/6.reorderList/reorder_list.py
"""
Reorder List

Given the head of a singly linked list: L0 → L1 → … → Ln-1 → Ln
Reorder it to: L0 → Ln → L1 → Ln-1 → L2 → Ln-2 → …

Time Complexity: O(n)
Space Complexity: O(1)
"""


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reorder_list(head):
    """
    Three steps: find middle, reverse second half, merge.
    TC: O(n), SC: O(1)
    """
    if not head or not head.next:
        return
    
    # Step 1: Find the middle
    slow = fast = head
    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next
    
    # Step 2: Reverse the second half
    second = slow.next
    slow.next = None  # Cut the list
    prev = None
    
    while second:
        next_temp = second.next
        second.next = prev
        prev = second
        second = next_temp
    
    second = prev
    
    # Step 3: Merge two halves
    first = head
    while second:
        tmp1, tmp2 = first.next, second.next
        first.next = second
        second.next = tmp1
        first, second = tmp1, tmp2


def print_list(head):
    """Helper to print linked list."""
    values = []
    while head:
        values.append(str(head.val))
        head = head.next
    return " -> ".join(values)


# Test case
if __name__ == "__main__":
    # Create list: 1 -> 2 -> 3 -> 4 -> 5
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    
    print(f"Original: {print_list(head)}")
    reorder_list(head)
    print(f"Reordered: {print_list(head)}")  # 1 -> 5 -> 2 -> 4 -> 3
